hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/utils/ActiveInstant.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.client.utils;
+
+import org.apache.hudi.common.table.HoodieTableMetaClient;
+import org.apache.hudi.common.table.timeline.HoodieInstant;
+import org.apache.hudi.common.table.timeline.HoodieTimeline;
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.common.util.ValidationUtils;
+
+import javax.annotation.Nullable;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * A combination of instants covering action states: requested, inflight, completed.
+ */
+public class ActiveInstant implements Serializable, Comparable<ActiveInstant> {
+  private final HoodieInstant requested;
+  private final HoodieInstant inflight;
+  private final HoodieInstant completed;
+
+  /**
+   * The constructor.
+   */
+  protected ActiveInstant(@Nullable HoodieInstant requested, @Nullable HoodieInstant inflight, HoodieInstant completed) {
+    this.requested = requested;
+    this.inflight = inflight;
+    this.completed = completed;
+  }
+
+  public static ActiveInstant fromInstants(List<HoodieInstant> instants) {
+    ValidationUtils.checkArgument(instants.size() <= 3);
+    HoodieInstant requested = null;
+    HoodieInstant inflight = null;
+    HoodieInstant completed = null;
+    for (HoodieInstant instant : instants) {
+      if (instant.isRequested()) {
+        requested = instant;
+      } else if (instant.isInflight()) {
+        inflight = instant;
+      } else {
+        completed = instant;
+      }
+    }
+    return new ActiveInstant(requested, inflight, Objects.requireNonNull(completed));
+  }
+
+  public List<HoodieInstant> getPendingInstants() {
+    List<HoodieInstant> instants = new ArrayList<>(2);
+    if (this.requested != null) {
+      instants.add(this.requested);
+    }
+    if (this.inflight != null) {
+      instants.add(this.inflight);
+    }
+    return instants;
+  }
+
+  public HoodieInstant getCompleted() {
+    return completed;
+  }
+
+  public String getAction() {
+    return this.completed.getAction();
+  }
+
+  /**
+   * A COMPACTION action eventually becomes COMMIT when completed.
```

Comment:
something to think about is - whether we keep it "COMPACTION" with the new changes.

Reply:
Personally I'm +1 for keeping the action of `compaction` and `log_compaction` just as it is, this avoids many ambiguilties, but I kind of think it should be a separate topic, we need a discuss whether to reuse the action for all kinds of table services: compaction, log_compaction, clustering, etc.

Reply:
+1 we can decouple this.